File: src/config/xray/protocols/outbound_settings.rs
use serde::{Deserialize, Serialize};

use super::common::{FragmentObject, NoiseObject, ReverseTagObject, WireguardPeerObject};
use crate::config::xray::shared::{Address, DomainStrategy, Network};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsBlackhole {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response: Option<BlackholeResponse>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BlackholeResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsDns {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub network: Option<Network>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub address: Option<Address>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub non_ip_query: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub block_types: Option<Vec<i32>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsFreedom {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain_strategy: Option<DomainStrategy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub redirect: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fragment: Option<FragmentObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub noises: Option<Vec<NoiseObject>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub proxy_protocol: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ips_blocked: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsHttp {
    pub address: Address,
    pub port: u16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pass: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsHysteria {
    #[serde(flatten)]
    pub extra: std::collections::HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsLoopback {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub inbound_tag: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsShadowsocks {
    pub address: Address,
    pub port: u16,
    pub method: String,
    pub password: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uot: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uot_version: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsSocks {
    pub address: Address,
    pub port: u16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pass: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsTrojan {
    pub address: Address,
    pub port: u16,
    pub password: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsVless {
    pub address: Address,
    pub port: u16,
    pub uuid: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flow: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub encryption: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reverse: Option<ReverseTagObject>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsVmess {
    pub address: Address,
    pub port: u16,
    pub uuid: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alter_id: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutboundSettingsWireguard {
    pub secret_key: String,
    pub address: Vec<String>,
    pub peers: Vec<WireguardPeerObject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mtu: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reserved: Option<Vec<i32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workers: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain_strategy: Option<DomainStrategy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kernel_mode: Option<bool>,
}

File: docs/plan/types/protocols/outbound-settings.ts
import type { Address, DomainStrategy, Network, StringMap } from "../shared";

export interface FragmentObject {
  packets?: string;
  length?: string;
  interval?: string;
}

export interface NoiseObject {
  type?: "rand" | "str" | "hex" | "base64" | string;
  packet?: string;
  delay?: string;
}

export interface WireguardPeerObject {
  endpoint: string;
  publicKey: string;
  preSharedKey?: string;
  keepAlive?: number;
  allowedIPs?: string[];
}

export interface OutboundSettingsBlackhole {
  response?: {
    type?: "http" | "none";
  };
}

export interface OutboundSettingsDns {
  network?: Extract<Network, "tcp" | "udp">;
  address?: Address;
  port?: number;
  userLevel?: number;
  nonIPQuery?: "drop" | "skip" | "reject" | string;
  blockTypes?: number[];
}

export interface OutboundSettingsFreedom {
  domainStrategy?: DomainStrategy;
  redirect?: string;
  userLevel?: number;
  fragment?: FragmentObject;
  noises?: NoiseObject[];
  proxyProtocol?: number;
  ipsBlocked?: string[];
}

export interface OutboundSettingsHttp {
  address: Address;
  port: number;
  user?: string;
  pass?: string;
  level?: number;
  email?: string;
  headers?: StringMap;
}

export interface OutboundSettingsHysteria {
  version: 2;
  address: Address;
  port: number;
}

export interface OutboundSettingsLoopback {
  inboundTag: string;
}

export interface OutboundSettingsShadowsocks {
  email?: string;
  address: Address;
  port: number;
  method: string;
  password: string;
  uot?: boolean;
  UoTVersion?: 1 | 2;
  level?: number;
}

export interface OutboundSettingsSocks {
  address: Address;
  port: number;
  user?: string;
  pass?: string;
  level?: number;
  email?: string;
}

export interface OutboundSettingsTrojan {
  address: Address;
  port: number;
  password: string;
  email?: string;
  level?: number;
}

export interface OutboundSettingsVless {
  address: Address;
  port: number;
  id: string;
  encryption: "none";
  flow?: string;
  level?: number;
  reverse?: {
    tag?: string;
  };
}

export interface OutboundSettingsVmess {
  address: Address;
  port: number;
  id: string;
  security?: "aes-128-gcm" | "chacha20-poly1305" | "auto" | "none" | "zero";
  level?: number;
  experiments?: string;
}

export interface OutboundSettingsWireguard {
  secretKey: string;
  address?: string[];
  noKernelTun?: boolean;
  mtu?: number;
  reserved?: number[];
  workers?: number;
  peers: WireguardPeerObject[];
  domainStrategy?: Extract<DomainStrategy, "ForceIPv6v4" | "ForceIPv6" | "ForceIPv4v6" | "ForceIPv4" | "ForceIP">;
}
